## In Class Assignment 2
### Linear Regression, Random Forest, PCA with Linear, PCA with Random Forest

In [137]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [138]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Chapter8_PCA/newhousing.csv')

In [139]:
data.head()

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  basement  \
0  5250000  5500         3          2        1         1          0         1   
1  4480000  4040         3          1        2         1          0         0   
2  3570000  3640         2          1        1         1          0         0   
3  2870000  3040         2          1        1         0          0         0   
4  3570000  4500         2          1        1         0          0         0   

   hotwaterheating  airconditioning  parking  prefarea  semi-furnished  \
0                0                0        0         0               1   
1                0                0        1         0               1   
2                0                0        0         0               0   
3                0                0        0         0               0   
4                0                0        0         0               0   

   unfurnished  areaperbedroom   bbratio  
0            0     1833.333333  0.666667  
1            0     1346.666667  0.333333  
2            1     1820.000000  0.500000  
3            1     1520.000000  0.500000  
4            0     2250.000000  0.500000

In [140]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   price            545 non-null    int64  
 1   area             545 non-null    int64  
 2   bedrooms         545 non-null    int64  
 3   bathrooms        545 non-null    int64  
 4   stories          545 non-null    int64  
 5   mainroad         545 non-null    int64  
 6   guestroom        545 non-null    int64  
 7   basement         545 non-null    int64  
 8   hotwaterheating  545 non-null    int64  
 9   airconditioning  545 non-null    int64  
 10  parking          545 non-null    int64  
 11  prefarea         545 non-null    int64  
 12  semi-furnished   545 non-null    int64  
 13  unfurnished      545 non-null    int64  
 14  areaperbedroom   545 non-null    float64
 15  bbratio          545 non-null    float64
dtypes: float64(2), int64(14)
memory usage: 68.2 KB


In [141]:
x=data.drop('price', axis=1)
y=data[['price']]

In [142]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=.20, random_state=42)

## LinearRegression


In [143]:
from sklearn.linear_model import LinearRegression
regressor=LinearRegression()
Linear_regressor_model=regressor.fit(x_train, y_train)

In [144]:
Linear_regressor_model.score(x_test, y_test)

0.6789132973714737

## Random Forest Regressor

In [145]:
from sklearn.ensemble import RandomForestRegressor
RFC=RandomForestRegressor(n_estimators=100)
RFC.fit(x_train, y_train)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


RandomForestRegressor()

In [146]:
RFC.score(x_test, y_test)

0.6860780865168605

## PCA with Linear Regression

In [147]:
from sklearn.preprocessing import StandardScaler
sc_x=StandardScaler()
sc_y = StandardScaler()
x=sc_x.fit_transform(x)
y=sc_y.fit_transform(y)

In [148]:
from sklearn.decomposition import PCA
pca = PCA(n_components=2)
pca.fit(x)
x_pca = pca.transform(x)

In [149]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size=.20, random_state=42)

In [150]:
regressor=LinearRegression()
Linear_regressor_model=regressor.fit(x_train, y_train)

In [151]:
Linear_regressor_model.score(x_test, y_test)

0.6444383990601952

## PCA with Random Forest

In [152]:
RFC=RandomForestRegressor(n_estimators=100)
RFC.fit(x_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


RandomForestRegressor()

In [153]:
RFC.score(x_test, y_test)

0.5985472830901475

## Conclusion
### Normal Random Forest Regressor performed best at 68%